### PR TITLE
[`pydoclint`] Implement `docstring-missing-exception` and `docstring-extraneous-exception` (`DOC501`, `DOC502`)

### PR DESCRIPTION
**Pull request by augustelalande**
## Summary

These are the first rules implemented as part of #458, but I plan to implement more.

Specifically, this implements `docstring-missing-exception` which checks for raised exceptions not documented in the docstring, and `docstring-extraneous-exception` which checks for exceptions in the docstring not present in the body.

## Test Plan

Test fixtures added for both google and numpy style.

### CODE REVIEW

**Comment by cr1901** [comment]
If this PR is accepted, will this preclude `pydoclint` (`darglint` successor but not really backwards-compat) support in the future? While `pydoclint` is the tool I personally use, it seems since yesterday, there are few thumbs-up asking [about using `pydoclint`](https://github.com/astral-sh/ruff/issues/458#issuecomment-2122773827) as a base instead.

**Comment by augustelalande** [comment]
I'm not opposed to using pydocstyle instead, but I'll wait until I get a review from the ruff team to change anything.

**Review by charliermarsh — APPROVED** [review]
Nice, this looks great. I'll make a few tweaks and merge today.

**Comment by augustelalande** [comment]
@charliermarsh where do you stand on darglint vs pydoclint as mentioned above

**Comment by charliermarsh** [comment]
Honestly, finding it hard to have a strong opinion on it... What do you think?

Absent other opinions, let's run with `pydoclint` given that it's actively developed? (And re-code these as `DOC501` and `DOC502`?)

**Comment by augustelalande** [comment]
Sounds reasonable to me

File: crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs:60]: What if this is like, `raise foo.FasterThanLightError from exc`? Would / should the docstring contain `FasterThanLightError` or `foo.FasterThanLightError`?

File: crates/ruff_linter/src/rules/pydocstyle/helpers.rs
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/pydocstyle/helpers.rs:99]: I'd probably replace this `break` control flow with an early `return`, since this method just returns `section_contexts` anyway.

File: crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs:271]: This implies function-only, right?

File: crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs:172]: Can you document this and the above to give an example of the expected structure that it's parsing?

File: crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs:160]: Could this return `Vec<&str>`?

File: crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs:225]: Per my comment above, I wonder if we need to match on attributes here too. What do you think?

File: crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs:228]: I wonder if we instead want to store the resolved name... Like if the user does:

```python
from foo import Bar as Baz

def func():
   """Raises: Bar"""
   raise Baz()
```

Should that be considered ok, or no?

File: crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs
Comment (augustelalande) [reply crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs:271]: Yes. At least for now since exceptions can't be raised by classes AFAIK.

File: crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs
Comment (augustelalande) [reply crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs:60]: Hard to say. I found nothing in the google style guide or numpy style guide. Per your other comments, I'll try to make it agnostic.

File: crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs
Comment (augustelalande) [reply crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs:244]: How would you resolve the qualified name from the docstring? Say we have 
```python
import foo

def func():
   """Raises: Bar"""
   raise foo.Bar()
```
Can we even resolve that this is correct?

Comment (augustelalande) [reply crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs]: Ok so I made it so that any suffix of the fully qualified name is allowed in the docstring

File: crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs
Comment (T-256) [inline crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs:75]:
```suggestion
/// ## What it does
```

File: crates/ruff_linter/src/checkers/ast/analyze/definitions.rs
Comment (ddelange) [inline crates/ruff_linter/src/checkers/ast/analyze/definitions.rs:328]: 1. will the pydoclint rules get their own settings section, or will pydoclint re-use the existing [`pydocstyle.convention`](https://docs.astral.sh/ruff/faq/#does-ruff-support-numpy-or-google-style-docstrings) setting?
2. if DOC gets a `pydoclint` settings section, will the user have to duplicate `convention` directive?
3. will `convention` be a required config, or what will be the default?
4. will the pydoclint rules support `pep257` convention just like pydocstyle?

Comment (charliermarsh) [reply crates/ruff_linter/src/checkers/ast/analyze/definitions.rs:328]: I think we'll just reuse the `convention` setting from `pydocstyle` for now and document it as such in the `pydoclint` rules. Eventually, we'll want to recategorize all of these under a single `docstrings` section, so it'll be unified at some point in the future.

Comment (augustelalande) [reply crates/ruff_linter/src/checkers/ast/analyze/definitions.rs:328]: 4. Yes we will eventually add support for `pep257` and `sphinx` style docstrings

File: crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs:147]: I changed this to return `None` when there's no section, so we didn't have to `unwrap` the `raised_exceptions_range`. Feel free to revert in future PRs if that gets in the way, though I think something _like_ that is useful to get rid of the unwrap.

**Comment by T-256** [comment]
Do we need to update title and description of #458 to keep tracking future implementations for this plugin?

**Comment by augustelalande** [comment]
I'm thinking to open a new issue to track `pydoclint` and supersede #458

**Pull request merged**
